src/aliased_buffer.h

```diff
@@ -26,27 +29,69 @@ namespace node {
  * The encapsulation herein provides a placeholder where such writes can be
  * observed. Any notification APIs will be left as a future exercise.
  */
+
+enum class AliasedBufferType {
+  kNew,
+  kCopy,
+  kShared,
+  kAssigned,
+  kToBeDeserialized,
+  kDeserialized
+};
+
 template <class NativeT,
           class V8T,
           // SFINAE NativeT to be scalar
-          typename = std::enable_if_t<std::is_scalar<NativeT>::value>>
+          typename Trait = std::enable_if_t<std::is_scalar<NativeT>::value>>
 class AliasedBufferBase {
  public:
-  AliasedBufferBase(v8::Isolate* isolate, const size_t count)
-      : isolate_(isolate), count_(count), byte_offset_(0) {
-    CHECK_GT(count, 0);
+  static std::unordered_set<AliasedBufferBase*> buffers;
+
+  inline AliasedBufferBase(v8::Isolate* isolate,
+                           const size_t count,
+                           const AliasedBufferInfo* info = nullptr)
+      : isolate_(isolate),
+        count_(count),
+        byte_offset_(0),
+        type_(AliasedBufferType::kNew) {
+    buffers.insert(this);
     const v8::HandleScope handle_scope(isolate_);
-    const size_t size_in_bytes =
-        MultiplyWithOverflowCheck(sizeof(NativeT), count);
+    if (info == nullptr) {
+      CHECK_GT(count, 0);
+      const size_t size_in_bytes =
+          MultiplyWithOverflowCheck(sizeof(NativeT), count);
+
+      // allocate v8 ArrayBuffer
+      v8::Local<v8::ArrayBuffer> ab =
+          v8::ArrayBuffer::New(isolate_, size_in_bytes);
+      buffer_ = static_cast<NativeT*>(ab->GetBackingStore()->Data());
+
+      // allocate v8 TypedArray
+      v8::Local<V8T> js_array = V8T::New(ab, byte_offset_, count);
+      js_array_ = v8::Global<V8T>(isolate, js_array);
+    } else {
+      // To be deserialized later when context is available
+      type_ = AliasedBufferType::kToBeDeserialized;
+      info_ = info;
+      buffer_ = nullptr;
+    }
+  }
 
-    // allocate v8 ArrayBuffer
-    v8::Local<v8::ArrayBuffer> ab = v8::ArrayBuffer::New(
-        isolate_, size_in_bytes);
-    buffer_ = static_cast<NativeT*>(ab->GetBackingStore()->Data());
+  AliasedBufferInfo Serialize(v8::Local<v8::Context> context,
+                              v8::SnapshotCreator* creator) {
+    return creator->AddData(context, GetJSArray());
+  }
 
-    // allocate v8 TypedArray
-    v8::Local<V8T> js_array = V8T::New(ab, byte_offset_, count);
-    js_array_ = v8::Global<V8T>(isolate, js_array);
+  inline void Deserialize(v8::Local<v8::Context> context) {
+    DCHECK_EQ(type_, AliasedBufferType::kToBeDeserialized);
+    v8::Local<V8T> arr =
+        context->GetDataFromSnapshotOnce<V8T>(*info_).ToLocalChecked();
+    DCHECK_EQ(count_, arr->Length());
+    byte_offset_ = arr->ByteOffset();
+    buffer_ = static_cast<NativeT*>(arr->Buffer()->GetBackingStore()->Data());
```

Comment:
This needs to take `byte_offset_` into account, though?

Reply:
I think we should turn this into a `CHECK_EQ(arr->ByteOffset(), 0)` - AliasedBuffers as views don't really need to be deserialized, instead it should be fine to just recreate them over the serialized backing buffer (what `CreateView()` does below).

Reply:
I’m not really a fan how that means that we introduce different types of `AliasedBuffer`s now, which need to be handled diffferently during serialization, and enforcing a byte offset of 0 sounds like it might be limiting in the future…

Reply:
Why would that be limiting? It only means that there are two types of AliasedBuffers (the distinction already exists), see :

1) AliasedBuffer that own their buffer and aren't views
2) AliasedBuffer that are views over existing AliasedBuffer (see the TODO in the header)

If we complete that TODO, then AliasedBuffer should assert the byte offset to be 0 anyway.

Reply:
I opened https://github.com/nodejs/node/pull/33142 to clarify the ownership of these buffers, considering the distinction already exists.

Reply:
I removed this from this PR as this is not really necessary for snapshotting the aliased buffers. Let's leave it to https://github.com/nodejs/node/pull/33142 (I still think this currently abstracts over both ArrayBuffer and TypedArray which is confusing though)